### PR TITLE
BUG: cells are missing in the excel file when exporting excel using xlsxwriter with option constant_memory set to True (#15392)

### LINKED ISSUE #15392: DataFrame.to_excel with xlsxwriter and constant_memory makes most of the cells empty
#### Problem description

Hello,

When we export data frames to excel using `xlsxwriter` with the option `constant_memory` set to `True`, most of the cells are empty.

Thanks a lot

#### Code Sample

```python
import seaborn.apionly as sns
import pandas as pd
iris = sns.load_dataset('iris')
with pd.ExcelWriter('constant_memory_false.xlsx', engine='xlsxwriter', options=dict(constant_memory=False)) as xlw:
    iris.to_excel(xlw)  # Open it: this is OK
with pd.ExcelWriter('constant_memory_true.xlsx', engine='xlsxwriter', options=dict(constant_memory=True)) as xlw:
    iris.to_excel(xlw)  # Open it: most of the data are missing
```

#### Output of ``pd.show_versions()``

```python
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 3.13.0-87-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.19.2
nose: 1.3.7
pip: 9.0.1
setuptools: 34.1.1
Cython: 0.25.2
numpy: 1.12.0
scipy: 0.18.1
statsmodels: 0.6.1
xarray: None
IPython: 5.2.2
sphinx: 1.5.2
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: 1.2.0
tables: 3.3.0
numexpr: 2.6.2
matplotlib: 2.0.0
openpyxl: None
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: None
bs4: 4.5.3
html5lib: 0.9999999
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: 2.6.2 (dt dec pq3 ext lo64)
jinja2: 2.9.5
boto: None
pandas_datareader: None
```

### PR DESCRIPTION
**Pull request by KenTsui**
- [x] closes #15392
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.25.1.rst [hunk 1]
```diff
@@ -24,6 +24,9 @@ Other enhancements
 
 Bug fixes
 ~~~~~~~~~
+I/O
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.1.rst]: there is an IO section already, move this there

File: pandas/core/generic.py [hunk 2]
```diff
@@ -2230,6 +2230,15 @@ def to_excel(
     ):
         df = self if isinstance(self, ABCDataFrame) else self.to_frame()
 
+        from pandas.io.excel._xlsxwriter import _XlsxWriter
```
Comment (jreback) [inline pandas/core/generic.py]: this is not the appropriate place for this, it needs to be in _XlsxWriter

File: pandas/io/formats/excel.py [hunk 3]
```diff
@@ -635,35 +638,45 @@ def _format_hierarchical_rows(self):
...
+                    colidx += 1
+
+                for cell in self._generate_body(colidx):
+                    yield cell
 
             else:
-                # Format hierarchical rows with non-merged values.
-                for indexcolvals in zip(*self.df.index):
-                    for idx, indexcolval in enumerate(indexcolvals):
+                _, ncol = self.df.shape
```
Comment (jreback) [inline pandas/io/formats/excel.py]: what are all of these changes about?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 2 above)
Comment (KenTsui) [reply pandas/core/generic.py]: Need to disable merge_cell based on 1) if it is using _XlsxWriter and 2) if it has set constant_memory True. In _XlsxWriter, you cannot control `merge_cell`. I will try to move it this piece of logic to pandas\io\formats\excel.py

File: pandas/io/formats/excel.py (hunk 3 above)
Comment (KenTsui) [reply pandas/io/formats/excel.py]: The previous code will generate `ExcelCell` of `df.index` column by column and then `ExcelCell` of `df.values` column by column. Now generate the first row of `ExcelCell` of `df.index` and `df.values`, then the second row and so on.

PS: `gcolidx` is revised to `colidx` for consistency only, as I use `rowidx` to represent row index.

File: pandas/core/generic.py (hunk 2 above)
Comment (gfyoung) [reply pandas/core/generic.py]: I think placing it in `excel.py` is a better place than `generic.py` for starters.  We can then evaluate feasibility of moving `_XlsxWriter` if necessary, though your points are well-taken.

File: pandas/io/formats/excel.py (hunk 3 above)
Comment (gfyoung) [reply pandas/io/formats/excel.py]: The approach is consistent with what was suggested in the original issue:

https://github.com/pandas-dev/pandas/issues/15392#issuecomment-279699881

File: pandas/core/generic.py (hunk 2 above)
Comment (KenTsui) [reply pandas/core/generic.py]: Now the testing if using `_XlsxWriter` and `constant_memory==True` is moved to `excel.py`. 
And by referencing the solution of an [issue](https://github.com/jmcnamara/XlsxWriter/issues/292) of `xlsxwriter`, there is a perk in the updated code that we can keep merged cells even if the two aforementioned conditions are present.

File: doc/source/whatsnew/v0.25.1.rst [hunk 4]
```diff
@@ -106,7 +106,7 @@ MultiIndex
 I/O
 ^^^
 
--
+- Bug in :func:`DataFrame.to_excel()` where cells are missing in the excel file when exporting excel using ``xlsxwriter`` with option ``constant_memory`` set to ``True`` ((:issue:`15392`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.25.1.rst]: Don't need the parens after to_excel.

Mismatched parens in the :issue:.

File: pandas/io/formats/excel.py [hunk 5]
```diff
@@ -466,14 +467,13 @@ def _format_header_mi(self):
...
         if self.merge_cells:
-            # Format multi-index as a merged cells.
-            for lnum in range(len(level_lengths)):
-                name = columns.names[lnum]
-                yield ExcelCell(lnum, coloffset, name, self.header_style)
-
             for lnum, (spans, levels, level_codes) in enumerate(
                 zip(level_lengths, columns.levels, columns.codes)
             ):
+                # Format multi-index as a merged cells.
```
Comment (TomAugspurger) [inline pandas/io/formats/excel.py]: Are these changes related to the constant memory issue?

Comment (KenTsui) [reply pandas/io/formats/excel.py]: yes, the constant memory issue is about the order of generating ExcelCell. So sometimes simply reorder the yield sequence can solve the problem.

**Comment by WillAyd** [comment]
I think all of this logic should be contained within the `write_cells` method of pandas.io.excel._xlsxwriter.py - does moving it there make things easier?

**Comment by KenTsui** [comment]
> I think all of this logic should be contained within the `write_cells` method of pandas.io.excel._xlsxwriter.py - does moving it there make things easier?

It seems not quite possible as the sequence of generating cells, which is the crux of this bug, is not controlled by `writer_cells`.

BTW, I am stuck on how to debug using azure pipelines.  Does anyone have some good resources to learn about that?

**Comment by WillAyd** [comment]
@KenTsui can you merge master? As far as log failures go if you click on next to failing logs you should be able to click through and find test failure output. Should match what you would see locally in a lot of cases

**Comment by WillAyd** [comment]
Closing as stale but ping if you'd like to pick back up
